qmf/src/plugins/messageservices/imap/imapstrategy.cpp

```diff
@@ -155,13 +155,14 @@ bool updateMessagesMetaData(ImapStrategyContextBase *context,
                                    const QMailMessageKey &storedKey, 
                                    const QMailMessageKey &unseenKey, 
                                    const QMailMessageKey &seenKey,
+                                   const QMailMessageKey &flaggedAsDeleted,
                                    const QMailMessageKey &flaggedKey,
                                    const QMailMessageKey &unreadElsewhereKey,
                                    const QMailMessageKey &importantElsewhereKey,
                                    const QMailMessageKey &unavailableKey)
 {
     bool result = true;
-    QMailMessageKey reportedKey(seenKey | unseenKey);
+    QMailMessageKey reportedKey((seenKey | unseenKey)  & ~flaggedAsDeleted);
```

Comment:
See line 177 (// Restore any messages thought to be unavailable that the server now reports), what is passed here is the entire reported list seen + unseen, so those include removed messages if they were not expunged, if we don't filter them out at line 177 they will be marked as undeleted locally(and flag will be sent back to server in next sync).